cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ControlDag.java

```diff
@@ -0,0 +1,301 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.planner;
+
+import co.cask.cdap.etl.proto.Connection;
+import com.google.common.base.Joiner;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multiset;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.UUID;
+
+/**
+ * A DAG (directed acyclic graph) where edges represent a happens-before relationship.
+ * In these types of scenarios, certain edges may be redundant and can be removed.
+ * This simplifies the dag into something that is much easier to work with if it needs
+ * to be used as a fork-join type of dag for workflow execution.
+ */
+public class ControlDag extends Dag {
+  private static final Set<String> EMPTY = ImmutableSet.of();
+  private final Multiset<String> nodeVisits;
+
+  public ControlDag(Collection<Connection> connections) {
+    super(connections);
+    this.nodeVisits = HashMultiset.create();
+  }
+
+  /**
+   * Record that this node was visited.
+   *
+   * @param node node that was visited
+   * @return the number of times this node was visited, including the visit from this call
+   */
+  public int visit(String node) {
+    nodeVisits.add(node);
+    return nodeVisits.count(node);
+  }
+
+  /**
+   * Resets the number of times each node was visited back to 0.
+   */
+  public void resetVisitCounts() {
+    nodeVisits.clear();
+  }
+
+  /**
+   * Flattens the control dag to remove connections between branches of different forks, which would
+   * make the dag unusable in pure fork-join workflows.
+   *
+   * For example the following dag is not a fork-join dag:
+   *
+   *      |--> n2 -------|
+   *      |              |--> n5
+   *      |--> n3 -------|
+   *  n1--|           |
+   *      |           v
+   *      |--> n4 --> n6
+   *
+   *  There are many ways to turn this a fork-join while still respecting all happens-before relationships,
+   *  but for simplicity we'll use an algorithm that doesn't have any nested forks and will turn the above into:
+   *
+   *       |--> n2 --|
+   *       |         |               |--> n5 --|
+   *  n1 --|--> n3 --|--> n2.n3.n4 --|         |--> n5.n6
+   *       |         |               |--> n6 --|
+   *       |--> n4 --|
+   *
```

Comment:
this class deals with dags where each edge indicates a happens-before relationship and is used to transform it into something our workflow system allows.  Our workflow system does not allow you to express the first dag.

The dags are most definitely not the same, but every happens-before relationship should be maintained.  In this case, n5 still happens after n2 and n3.  There is just an additional requirement that it also happens before n4.  Similarly, n6 still happens after n3 and n4, there is just an additional requirement that it also happens before n2.